account_id.go

```diff
@@ -39,16 +66,59 @@ func AccountIDFromSolidityAddress(s string) (AccountID, error) {
 	}
 
 	return AccountID{
-		Shard:   shard,
-		Realm:   realm,
-		Account: account,
+		Shard:    shard,
+		Realm:    realm,
+		Account:  account,
+		Checksum: nil,
+		Network:  nil,
 	}, nil
 }
 
+func (id *AccountID) GetNetworkFromChecksum() error {
+	if id.Checksum == nil {
+		return errors.New("Checksum is missing.")
+	}
+	var checksum parseAddressResult
+	var err error
+
+	var networkNames = []NetworkName{
+		Mainnet,
+		Testnet,
+		Previewnet,
+	}
+
+	var network NetworkName
+	for _, name := range networkNames {
+		checksum, err = checksumParseAddress(name.Network(), fmt.Sprintf("%d.%d.%d-%s", id.Shard, id.Realm, id.Account, *id.Checksum))
+		if err != nil {
+			return err
+		}
+		if checksum.status != 1 {
+			network = name
+			break
+		}
+	}
+
+	id.Network = &network
+	return nil
+}
+
+func AccountIDValidateNetworkOnIDs(id AccountID, other AccountID) error {
+	if !id.isZero() && !other.isZero() && id.Network != nil && other.Network != nil && *id.Network != *other.Network {
+		return errNetworkMismatch
+	}
+
+	return nil
+}
+
 // String returns the string representation of an AccountID in
 // `Shard.Realm.Account` (for example "0.0.3")
 func (id AccountID) String() string {
-	return fmt.Sprintf("%d.%d.%d", id.Shard, id.Realm, id.Account)
+	checksum, err := checksumParseAddress("", fmt.Sprintf("%d.%d.%d", id.Shard, id.Realm, id.Account))
```

Comment:
We should use the checksum already on the ID; checksums should be calculated when the ID is created.